### ISSUE TITLE
proof of concept of runtime heatmap histogram creation

### ISSUE THREAD

**Issue by carns** [darshan-hpc/darshan#468]
The current pydarshan-job-summary heatmap is generated using DXT trace data, which is not available in default production deployment logs.  We need to be able to generate equivalent data in the default Darshan runtime mode.

The purpose of this ticket is to demonstrate that we can do so with minimal runtime instrumentation overhead.  A later ticket would track the work to finalize the implementation and corresponding log format.

From #454 discussion:

```
The heat map uses an N x M array, where N is the number of x-bins used (default is 200), and M is the number of ranks. The number of x-bins is arbitrarily set to 200, which is probably more than necessary for logs with shorter run times, and too few for logs with longer run times (say 1000 seconds). I'm not sure if it could be setup to choose its own bin count, but that would probably be ideal so we don't lose too much info on longer runs.

The width of each bin is calculated from the total run time and the number of bins, at the moment. Obviously the run time wouldn't be available while the log is being generated, though.
```

We would probably record this in a new module with it's own record format that isn't tied to file names.  There will be a fixed number of bins allocated per rank at startup, with a fixed initial bin width (e.g., .5 seconds).  Once the bins are all populated, or within a given threshold, then the bin width value could be doubled and the bins themselves combined/condensed to be half-populated.

At any given time when the program exits, it will be the case that some bins are un-populated.  The parser or analysis tool can ignore any bins such that (bin_width * bin_number) > job_runtime, and/or we could mark the last valid one at shutdown.

If we can find a better algorithm (that doesn't require a bulk condense step, that would be great.  We can measure the performance of the above basic approach to see where we stand before we get too deep, though.

Bins probably need read and write counters (int64).  We can proportionally distribute bytes in cases where accesses span bin boundaries.

**Comment by tylerjereddy** [comment]
Perhaps more on the implementation detail side, but if this is to be a new module with new infrastructure surrounding it, it would be great if the binary data that gets stored in a darshan log file for this module could be read back into memory as a single contiguous C buffer that spans all the "records" (rows). I believe that would provide the Python bindings with extremely fast access to the information vs. repeated C interface calls to retrieve the rows/records of data. Related to #451.

**Comment by carns** [comment]
This issue was resolved in #548
